src/common/device/AndroidDevice.ts

```diff
@@ -103,18 +107,95 @@ export class AndroidDevice implements BaseDevice {
-    ): Promise<void> {
-        await AndroidUtils.launchAppInBootedEmulator(this.port, target, appBundlePath, targetAppArguments, this.logger);
+    public async launchApp(target: string, appBundlePath?: string, launchArguments?: LaunchArgument[]): Promise<void> {
+        await AndroidUtils.launchAppInBootedEmulator(this.port, target, appBundlePath, launchArguments, this.logger);
+    }
+
+    /**
+     * Checks to see if a certificate is already installed on the device.
+     *
+     * @param certData An SSLCertificateData object containing the certificate data.
+     * @returns A boolean indicating if a certificate is already installed on the device or not.
+     */
+    public async isCertInstalled(certData: SSLCertificateData): Promise<boolean> {
+        // For Android emulators, certificates are installed in files named as their subject hash.
+        // We can query the device to see if it has the file with the name as the subject hash of the
+        // provided certificate. If so then we can say that the cert is already installed.
+        const subjectHash = CryptoUtils.getSubjectHashOld(certData);
+        const fileName = `${subjectHash}.0`; // this is the special file name
+
+        try {
+            // start adb as root so that we can query specific folder location
+            await AndroidUtils.executeAdbCommand('root', this.port, this.logger);
+
+            // see if the file exists already
+            const result = await AndroidUtils.executeAdbCommand(
+                'shell "ls /data/misc/user/0/cacerts-added"',
+                this.port,
+                this.logger
+            );
+            return result.includes(fileName);
+        } catch (error) {
+            // If file doesn't exist then an error is thrown on some versions of adb.
+            // Also determining if cert is already installed is a best effort so if any errors
+            // occur we will continue and assume that cert is not installed.
+            this.logger?.warn(error);
+        }
+        return false;
+    }
+
+    /**
+     * Installs a certificate on the device.
+     *
+     * @param certData An SSLCertificateData object containing the certificate data.
+     */
+    public async installCert(certData: SSLCertificateData): Promise<void> {
+        // For Android emulators, we need to save the PEM certificate in a file named as the cert subject hash.
+        const subjectHash = CryptoUtils.getSubjectHashOld(certData);
+        const pemContent = certData.pemCertificate ?? CryptoUtils.derToPem(certData.derCertificate);
+        const fileName = `${subjectHash}.0`; // this is the special file name
+        const certFilePath = path.join(os.tmpdir(), fileName);
+        fs.writeFileSync(certFilePath, pemContent);
+
+        // We then need to push the file to the emulator (needs to be root-mountable).
+        await AndroidUtils.mountAsRootWritableSystem(this.id, this.logger); // boot with writable system access
```

Comment:
Will adding certs in this way require that users then boot their emulators with `-writable-system`, in order for the changes to persist?

Reply:
No It won't. After the cert is installed, if the user later on starts the device regularly (for example using Android Studio) the cert will still be there.